# Pandas DataFrames 101 - Learning Notebook

----

<img src="https://miro.medium.com/v2/resize:fit:1400/1*Q55X1gfHjKi1knx96UVdZQ.png" alt="Description" width="450">

## Contents

In this notebook we will cover the following:

1. **Introduction to DataFrames**
    1. What is a DataFrame?
    2. Reading data into a dataframe
2. **Previewing and describing a DataFrame**
    1. Common methods for a quick overview
3. **Selecting and filtering data**
    1. Selecting rows and columns
    2. Using masks
4. **Analyzing and aggregating data**
    1. Analysing numerical data
    2. Analyzing categorical data
    3. Using `.group_by()`

## Imports

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
import pandas as pd
import os

# 1. Introduction to DataFrames

## What is a DataFrame

A 2D, potentially heterogenous, tabular structure. The documentation on DataFrame is available on the `pandas.DataFrame` [documentation page](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html).

Let's start by reading some dataset on heart disease ([source](https://archive.ics.uci.edu/dataset/45/heart+disease)). 

## Reading data into a DataFrame

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/Hospital-Da-Luz-Learning-Health/MLCatolica24/main/Aula%203%20-%20Data%20Analysis%20%26%20Statistics/Pandas%20DataFrames%20101/data/heart_disease_cleaned.csv')
df = df.set_index('id')

# 2. Previewing and describing a DataFrame

## Common methods for a quick overview

Let's take a sneak peak at this data:

In [4]:
df.head()
#df.tail(3)
df.shape

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
192,43,male,asymptomatic,132,247,1,left ventricular hypertrophy,143,1,0.1,fat,NaN,reversible defect,>50%
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%
0,63,male,typical angina,145,233,1,left ventricular hypertrophy,150,0,2.3,downsloping,0.0,fixed defect,<50%
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
284,61,male,asymptomatic,148,203,0,normal,161,0,0.0,upsloping,1.0,reversible defect,>50%


(303, 14)

In [5]:
df.describe()

,age,trestbps,chol,fbs,thalach,exang,oldpeak,ca
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,299.000000
mean,54.438944,131.689769,246.693069,0.148515,149.607261,0.326733,1.039604,0.672241
std,9.038662,17.599748,51.776918,0.356198,22.875003,0.469794,1.161075,0.937438
min,29.000000,94.000000,126.000000,0.000000,71.000000,0.000000,0.000000,0.000000
25%,48.000000,120.000000,211.000000,0.000000,133.500000,0.000000,0.000000,0.000000
50%,56.000000,130.000000,241.000000,0.000000,153.000000,0.000000,0.800000,0.000000
75%,61.000000,140.000000,275.000000,0.000000,166.000000,1.000000,1.600000,1.000000
max,77.000000,200.000000,564.000000,1.000000,202.000000,1.000000,6.200000,3.000000


Here's what we know about this dataset:

- contains 303 rows (observations)
- 14 columns

And some documentation we found about each column:

1. **Age:** age in years
2. **Sex:** sex (1 = male; 0 = female)
3. **Cp:** chest pain type
   - Value 1: typical angina
   - Value 2: atypical angina
   - Value 3: non-anginal pain
   - Value 4: asymptomatic
4. **Trestbps:** resting blood pressure (in mm Hg on admission to the hospital)
5. **Chol:** serum cholesterol in mg/dl
6. **Fbs:** fasting blood sugar > 120 mg/dl (1 = true; 0 = false)
7. **Restecg:** resting electrocardiographic results
   - Value 0: normal
   - Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
   - Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
8. **Thalach:** maximum heart rate achieved
9. **Exang:** exercise induced angina (1 = yes; 0 = no)
10. **Oldpeak:** ST depression induced by exercise relative to rest
11. **Slope:** the slope of the peak exercise ST segment
    - Value 1: upsloping
    - Value 2: flat
    - Value 3: downsloping
12. **Ca:** number of major vessels (0-3) colored by flourosopy
13. **Thal:** 
    - 3 = normal
    - 6 = fixed defect
    - 7 = reversible defect
14. **Diagnosed:** diagnosis of heart disease (angiographic disease status)
    - Value 0: < 50% diameter narrowing
    - Value 1: > 50% diameter narrowing


# 3. Selecting and filtering data

## Selection rows and columns

<img src="media/df_filtering.png" width="900">

In [6]:
df['age'] # selects a column and return a pandas series

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

In [7]:
df.loc[:189,'age'] # does the same thing

id
192    43
189    69
Name: age, dtype: int64

We can access the index like so:

In [8]:
df.index

Index([192, 189,   0,  23, 284, 177,  89, 290, 127,  56,
       ...
       163, 169, 186, 244, 139,   9, 109, 185, 111,  36],
      dtype='int64', name='id', length=303)

And this works like a list, so let's see who is the patient in the 41st line

In [9]:
df.index[40]

235

> What is the `age` of the patient whose `id` is `235`?

In [10]:


# Expected output: 54

> What is the `slope` and `cp` of the patient whose `id` is `177`?

In [11]:


# Expected output:
    # slope             fat
    # cp       asymptomatic

> What is the `sex` and `thal` of patients `23`, `199` and `102`?

In [12]:


# Expected output:
    # 23     male   reversible defect
    # 199    male      normal
    # 102   female     normal

## Using Masks

<img src="https://images7.memedroid.com/images/UPLOADED669/640ef2a4662e1.jpeg" alt="Description" width="350">

Masks are great because they allow us to select only part of data you want

> Select the patients with `restecg` (resting electrocardiographic results) not normal

In [13]:


# Expected output:
    # id
    # 192     True
    # 189     True
    # 0       True
    # 23      True
    # 284     False
    #        ...  

In [14]:
mask = df['age'] > 55

In [15]:
mask

id
192    False
189     True
0       True
23      True
284     True
       ...  
9      False
109    False
185     True
111     True
36     False
Name: age, Length: 303, dtype: bool

What is a mask?
- A series with a bunch of Falses and some Trues
- The **True** values are for the rows where the condition above is ... well ... True!

<!-- ![](media/what.jpeg) -->
<img src="media/what.jpeg" width="400">

In [16]:
mask.value_counts()

age
True     152
False    151
Name: count, dtype: int64

and now we use this mask to select only the rows with `True`

In [17]:
df.loc[mask,:]

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%
0,63,male,typical angina,145,233,1,left ventricular hypertrophy,150,0,2.3,downsloping,0.0,fixed defect,<50%
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
284,61,male,asymptomatic,148,203,0,normal,161,0,0.0,upsloping,1.0,reversible defect,>50%
177,56,male,asymptomatic,132,184,0,left ventricular hypertrophy,105,1,2.1,fat,1.0,fixed defect,>50%
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
251,58,male,asymptomatic,146,218,0,normal,105,0,2.0,fat,1.0,reversible defect,>50%
163,58,female,asymptomatic,100,248,0,left ventricular hypertrophy,122,0,1.0,fat,0.0,normal,<50%
244,60,female,non-anginal pain,120,178,1,normal,96,0,0.0,upsloping,0.0,normal,<50%


In [18]:
# let's confirm the above
df.loc[mask,'age'].min()

56

# 4. Descriptive statistics and aggregating data

What type of data do we have?

In [19]:
df.dtypes

age            int64
sex           object
cp            object
trestbps       int64
chol           int64
fbs            int64
restecg       object
thalach        int64
exang          int64
oldpeak      float64
slope         object
ca           float64
thal          object
Diagnosed     object
dtype: object

## Analyzing numerical data - mean, median, max, min, describe

In [20]:
df['age'].head(3)

id
192    43
189    69
0      63
Name: age, dtype: int64

In [21]:
df['age'].mean()

54.43894389438944

In [22]:
df['age'].median()

56.0

In [23]:
df['age'].max()

77

In [24]:
df['age'].min()

29

[Here are all descriptive statistics you can easily compute from a Pandas series](https://pandas.pydata.org/pandas-docs/version/0.20.2/api.html#api-dataframe-stats)

You can also use `.describe()` to get the most common stats of a numerical column

In [25]:
df['age'].describe()

count    303.000000
mean      54.438944
std        9.038662
min       29.000000
25%       48.000000
50%       56.000000
75%       61.000000
max       77.000000
Name: age, dtype: float64

in case you want a better display, you can convert a pandas series to dataframe with `.to_frame()`

In [26]:
# in case you want a better display, you can convert a pandas series to dataframe 
df['age'].describe().to_frame()

,age
count,303.000000
mean,54.438944
std,9.038662
min,29.000000
25%,48.000000
50%,56.000000
75%,61.000000
max,77.000000


## Analyzing categorical data with

In [27]:
df.dtypes

age            int64
sex           object
cp            object
trestbps       int64
chol           int64
fbs            int64
restecg       object
thalach        int64
exang          int64
oldpeak      float64
slope         object
ca           float64
thal          object
Diagnosed     object
dtype: object

> How many males and females are there in the dataset?

In [28]:


# Expected output:
    # sex
    # male      206
    # female     97

> In percentage? And how do you round the percentages?

In [29]:


# Expected output:
    # sex
    # male      0.679868
    # female    0.320132

Can this proportion be a problem in the future?

----

I would like to look at the proportion of all categorical columns

In [30]:
df.select_dtypes('object')

,sex,cp,restecg,slope,thal,Diagnosed
id,,,,,,
192,male,asymptomatic,left ventricular hypertrophy,fat,reversible defect,>50%
189,male,non-anginal pain,left ventricular hypertrophy,fat,reversible defect,>50%
0,male,typical angina,left ventricular hypertrophy,downsloping,fixed defect,<50%
23,male,non-anginal pain,left ventricular hypertrophy,upsloping,reversible defect,>50%
284,male,asymptomatic,normal,upsloping,reversible defect,>50%
...,...,...,...,...,...,...
9,male,asymptomatic,left ventricular hypertrophy,downsloping,reversible defect,>50%
109,male,asymptomatic,normal,fat,reversible defect,>50%
185,female,atypical angina,normal,upsloping,normal,<50%


<div style="border-left: 6px solid #4CAF50; background-color: #f0fdf4; padding: 10px; border-radius: 5px;">
  <h5>📝 Exercise 1</h5>
  <p>Using a for cycle, show the proportion of each category for each categorical variable.</p>
</div>

One cool thing about this is that you can better analyze your data, check the following example:

In [31]:
# mask = df['age'] > 55
mask

id
192    False
189     True
0       True
23      True
284     True
       ...  
9      False
109    False
185     True
111     True
36     False
Name: age, Length: 303, dtype: bool

In [32]:
df.loc[mask,'Diagnosed'].value_counts(normalize=True)

Diagnosed
>50%    0.592105
<50%    0.407895
Name: proportion, dtype: float64

> Does this pattern maintain for people younger than 55?

In [33]:

# Expected output
    # Diagnosed
    # <50%    0.675497
    # >50%    0.324503

Very different!

We can also join together masks as well with logical operations

In [34]:
mask_age_above_55 = df['age'] > 55
mask_cp_asymptomatic = df['cp'] == 'asymptomatic'

In [35]:
df.loc[mask_age_above_55 & mask_cp_asymptomatic,:].head(3)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
284,61,male,asymptomatic,148,203,0,normal,161,0,0.0,upsloping,1.0,reversible defect,>50%
177,56,male,asymptomatic,132,184,0,left ventricular hypertrophy,105,1,2.1,fat,1.0,fixed defect,>50%
96,59,male,asymptomatic,110,239,0,left ventricular hypertrophy,142,1,1.2,fat,1.0,reversible defect,>50%


## Using .groupby()

`Groupby` is useful when you want to ask a question about the dataset, and you want the answer stratified by groups!

<p align="center">
  <img src="media/gorupby.png" alt="image credit"><br>
  <em>Image adapted from <a href="https://towardsdatascience.com/11-examples-to-master-pandas-groupby-function-86e0de574f38">https://towardsdatascience.com/11-examples-to-master-pandas-groupby-function-86e0de574f38</a></em>
</p>

Let's take a look at the following example:

> "I want the average age of my patients, but stratified by sex (i.e. for males and females separately)"

We've learned with `masks` that we can do stuff like this already, so...

<div style="border-left: 6px solid #4CAF50; background-color: #f0fdf4; padding: 10px; border-radius: 5px;">
  <h5>📝 Exercise 2</h5>
  <p>Compute the average age of the patients, stratified by sex (i.e. for males and females separately) using masks.</p>
</div>

Using `groupby` instead, we get:

In [36]:
df.groupby('sex')['age'].mean()

sex
female    55.721649
male      53.834951
Name: age, dtype: float64

<img src="https://i.imgflip.com/8v206r.jpg" width="400">


In [37]:
# we can "prettify" the previous results
df.groupby('sex')['age'].mean().to_frame().round(1)

,age
sex,
female,55.7
male,53.8


In [38]:
df.groupby('sex').value_counts().to_frame()

count
sex    age cp               trestbps chol fbs restecg                      thalach exang oldpeak slope       ca  thal              Diagnosed       
female 34  atypical angina  118      210  0   normal                       192     0     0.7     upsloping   0.0 normal            <50%           1
       35  asymptomatic     138      183  0   normal                       182     0     1.4     upsloping   0.0 normal            <50%           1
       37  non-anginal pain 120      215  0   normal                       170     0     0.0     upsloping   0.0 normal            <50%           1
       39  non-anginal pain 94       199  0   normal                       179     0     0.0     upsloping   0.0 normal            <50%           1
                            138      220  0   normal                       152     0     0.0     fat         0.0 normal            <50%           1
...                                                                                                                                             ...
male   70  asymptomatic     130      322  0   left ventricular hypertrophy 109     0     2.4     fat         3.0 normal            >50%           1
                            145      174  0   normal                       125     1     2.6     downsloping 0.0 reversible defect >50%           1
           atypical angina  156      245  0   left ventricular hypertrophy 143     0     0.0     upsloping   0.0 normal            <50%           1
           non-anginal pain 160      269  0   normal                       112     1     2.9     fat         1.0 reversible defect >50%           1
       77  asymptomatic     125      304  0   left ventricular hypertrophy 162     1     0.0     upsloping   3.0 normal            >50%           1

[297 rows x 1 columns]

## Let's start to answer a series of questions using `groupby`

In [39]:
df.head(2)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
192,43,male,asymptomatic,132,247,1,left ventricular hypertrophy,143,1,0.1,fat,NaN,reversible defect,>50%
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%


> What is the frequency of patients of each cp category

In [40]:
df.groupby('cp').size().to_frame('Number of Patients')

,Number of Patients
cp,
asymptomatic,144
atypical angina,50
non-anginal pain,86
typical angina,23


Notice how we didn't have to select a column here because you don't need a specific column to compute the size of a group 

<div style="border-left: 6px solid #4CAF50; background-color: #f0fdf4; padding: 10px; border-radius: 5px;">
  <h5>📝 Exercise 3</h5>
  <p>What is the distribution of <code>oldpeak</code> (numerical) by <code>sex</code> (category)?</p>
</div>

**Reminder**: This is how we've already learned to quickly check some main distribution statistics of a *numerical* column:

<div style="border-left: 6px solid #4CAF50; background-color: #f0fdf4; padding: 10px; border-radius: 5px;">
  <h5>📝 Exercise 4</h5>
  <p>What is the distribution of <code>slope</code> (category) by <code>cp</code> (category)?</p>
</div>

While it seems we have 3 columns in the result above, this is actually just a pandas series whose index has 2 levels.

In [41]:
df.groupby('cp')['slope'].value_counts().index

MultiIndex([(    'asymptomatic',         'fat'),
            (    'asymptomatic',   'upsloping'),
            (    'asymptomatic', 'downsloping'),
            ( 'atypical angina',   'upsloping'),
            ( 'atypical angina',         'fat'),
            ( 'atypical angina', 'downsloping'),
            ('non-anginal pain',   'upsloping'),
            ('non-anginal pain',         'fat'),
            ('non-anginal pain', 'downsloping'),
            (  'typical angina',         'fat'),
            (  'typical angina',   'upsloping'),
            (  'typical angina', 'downsloping')],
           names=['cp', 'slope'])

either way, the results are hard to understand at a glance, so let's "prettify" this

In [42]:
df.groupby('cp')['slope'].value_counts().to_frame('Count') # simply converted the pandas series to a dataframe

Count
cp               slope             
asymptomatic     fat             84
                 upsloping       49
                 downsloping     11
atypical angina  upsloping       36
                 fat             12
                 downsloping      2
non-anginal pain upsloping       48
                 fat             33
                 downsloping      5
typical angina   fat             11
                 upsloping        9
                 downsloping      3

As a dataframe it looks better already but still not good enough

In [43]:
df.groupby('cp')['slope'].value_counts().to_frame('Count').unstack(level='cp')

Count                                                
cp          asymptomatic atypical angina non-anginal pain typical angina
slope                                                                   
downsloping           11               2                5              3
fat                   84              12               33             11
upsloping             49              36               48              9

`unstack()` pivots one of the index levels and has made the table a lot easier to interpret. You can see the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.unstack.html) for a more clear explanation and examples